text/3660-crates-io-crate-deletions.md

```diff
@@ -0,0 +1,123 @@
+- Start Date: 2024-06-20
+- RFC PR: [rust-lang/rfcs#3660](https://github.com/rust-lang/rfcs/pull/3660)
+
+# Summary
+
+This RFC proposes a mechanism for crate authors to delete their crates from crates.io under certain conditions.
+
+
+# Motivation
+
+There are a variety of reasons why a crate author might want to delete a crate or version from crates.io:
+
+* You published something accidentally.
+* You wanted to test crates.io.
+* You published content you didn't intend to be public.
+* You want to rename a crate. (The only way to rename a package is to re-publish it under a new name)
+
+The current [crates.io usage policy](https://crates.io/policies) says:
+
+> Crate deletion by their owners is not possible to keep the registry as immutable as possible.
+
+This restriction makes sense for the majority of crates that have been around for a while and are actively used, but the above list of reasons shows that there are valid use cases for allowing crate authors to delete their crates without having to contact the crates.io team.
+
+To make this process easier for our users and to reduce the workload of the crates.io team dealing with such support requests, we propose to codify our current set of informal rules into a formal policy that allows crate authors to delete their crates themselves under certain conditions (see below).
+
+
+# Proposal
+
+We propose to allow crate authors to delete their **crates** from crates.io under the following conditions:
+
+* The crate has been published for less than 72 hours,
+* or if all the following conditions are met:
+  * The crate has a single owner,
+  * The crate has no reverse dependencies,
+  * The crate has been downloaded less than 100 times for each month it has been published.
+
+We also propose to allow crate authors to delete **versions** of their crates from crates.io under the following conditions:
+
+* The version has been published for less than 72 hours.
+
+These crate owner actions will be enabled by two new API endpoints:
```

Comment:
Should we now or in the future expose a cargo command for this?

Reply:
I guess that is up to the cargo team. npm has an `unpublish` command for this purpose. this RFC explicitly only targets crates.io, not the general crate registry contract, which could potentially be something for the "Future possibilities" section.

Reply:
If there's not a cargo command, it'll be used scarcely even when doing so is appropriate. imo a command should be included (whether it needs to be part of this RFC is not by purview).

Reply:
@jhpratt IMHO we can implement this incrementally. this RFC is mostly about providing the functionality in general, and doesn't go into too much detail on how exactly this will be implemented. once the backend work in crates.io supports it we can experiment with ways to expose the functionality to users. e.g. we can add buttons on the website, we can build a cargo plugin, we can build it into cargo itself, etc.

Reply:
>  e.g. we can add buttons on the website, we can build a cargo plugin, we can build it into cargo itself, etc.

This was where I was starting to go.  Using a web frontend makes it easier to have and tweak an appropriate notice and a "are you sure" confirmation (like Github's "enter the repo name" field).  

imo the way to keep the "provide but not specify" the functionality would be to not specify the backend but just to say that there will be a frontend controls for this.  At least having it as a Future possibility makes your intent with this clearer.

Reply:
I've updated the "Future possibilities" section to mention the web frontend and cargo subcommand ideas.

Reply:
the RFC is mainly focused on allowing deletions in general. I didn't want to conflate it with the details on how exactly the functionality would be exposed to the users since these seem like implementation details that the crates.io and cargo teams can decide upon later.

Reply:
Sorry, I'm still a bit confused. Why would we approve an RFC for functionality that nobody can use? If you don't want to go into the details of exactly what the interface should look like, shouldn't the RFC at least explicitly say "the crates.io web interface will have functionality to perform these actions"?

My understanding of "future possibilities" isn't for things that we *will do*. It is for things that will be considered as possibilities for how it could interact with other changes, or ways it could be extended with future RFCs. Putting the implementation in the "future" section would be like saying "we're going to add coroutines to Rust, but we're not going to tell you what the language looks like, we'll decide that later". That doesn't fit the spirit of RFCs as I understand them.